lib/model/model.go

```diff
@@ -1353,7 +1353,7 @@ func (m *Model) Request(deviceID protocol.DeviceID, folder, name string, offset
 			return protocol.ErrNoSuchFile
 		}
 
-		if err := readOffsetIntoBuf(folderFs, tempFn, offset, buf); err == nil {
+		if err := readOffsetIntoBuf(folderFs, tempFn, offset, buf); err == nil && scanner.Validate(buf, hash, weakHash) {
```

Comment:
That's a fairly expensive operation to hide at the tail end of an if condition. Maybe make it its own thing below the reading?